src/couplings.jl

```diff
@@ -0,0 +1,244 @@
+
+Partitions `x` into 3 disjoint subvectors.
+"""
+@inline partition(m::PartitionMask, x) = (transpose(m.A_1) * x, transpose(m.A_2) * x, transpose(m.A_3) * x)
+
+
+# CouplingLayer
+
+"""
+    CouplingLayer{B, M, F}(mask::M, θ::F)
+
+Implements a coupling-layer as defined in [1].
+
+# Examples
+```julia-repl
+julia> m = PartitionMask(3, [1], [2]) # <= going to use x[2] to parameterize transform of x[1]
+PartitionMask{SparseArrays.SparseMatrixCSC{Float64,Int64}}(
+  [1, 1]  =  1.0, 
+  [2, 1]  =  1.0, 
+  [3, 1]  =  1.0)
+
+julia> cl = CouplingLayer(Shift, m, identity) # <= will do `y[1:1] = x[1:1] + x[2:2]`;
+
+julia> x = [1., 2., 3.];
+
+julia> cl(x)
+3-element Array{Float64,1}:
+ 3.0
+ 2.0
+ 3.0
+```
+
+# References
+[1] Kobyzev, I., Prince, S., & Brubaker, M. A., Normalizing flows: introduction and ideas, CoRR, (),  (2019). 
+"""
+struct CouplingLayer{B, M, F} <: Bijector{1} where {B, M <: PartitionMask, F}
+    mask::M
+    θ::F
+end
+
+CouplingLayer(B, mask::M, θ::F) where {M, F} = CouplingLayer{B, M, F}(mask, θ)
+function CouplingLayer(B, θ, n::Int)
+    idx = Int(floor(n / 2))
+    return CouplingLayer(B, PartitionMask(n, 1:idx), θ)
+end
+
+function CouplingLayer(cl::CouplingLayer{B}, mask::PartitionMask) where {B}
+    return CouplingLayer(B, mask, cl.θ)
+end
+
+"Returns the constructor of the coupling law."
+coupling(cl::CouplingLayer{B}) where {B} = B
+
+"Returns the coupling law constructed from `x`."
+function couple(cl::CouplingLayer{B}, x::AbstractVector) where {B}
+    # partition vector using `cl.mask::PartitionMask`
+    x_1, x_2, x_3 = partition(cl.mask, x)
+
+    # construct bijector `B` using θ(x₂)
+    b = B(cl.θ(x_2))
+
+    return b
+end
+
+function (cl::CouplingLayer{B})(x::AbstractVector) where {B}
+    # partition vector using `cl.mask::PartitionMask`
+    x_1, x_2, x_3 = partition(cl.mask, x)
+
+    # construct bijector `B` using θ(x₂)
+    b = B(cl.θ(x_2))
+
+    # recombine the vector again using the `PartitionMask`
+    return combine(cl.mask, b(x_1), x_2, x_3)
+end
+function (cl::CouplingLayer{B})(x::AbstractMatrix) where {B}
+    return hcat([cl(x[:, i]) for i = 1:size(x, 2)]...)
+end
+
+
+function (icl::Inversed{<:CouplingLayer{B}})(y::AbstractVector) where {B}
+    cl = icl.orig
+    
+    y_1, y_2, y_3 = partition(cl.mask, y)
+
+    b = B(cl.θ(y_2))
+    ib = inv(b)
+
+    return combine(cl.mask, ib(y_1), y_2, y_3)
+end
+function (icl::Inversed{<:CouplingLayer{B}})(y::AbstractMatrix) where {B}
+    return hcat([icl(y[:, i]) for i = 1:size(y, 2)]...)
+end
+
+function logabsdetjac(cl::CouplingLayer{B}, x::AbstractVector) where {B}
+    x_1, x_2, x_3 = partition(cl.mask, x)
+    b = B(cl.θ(x_2))
+
+    # `B` might be 0-dim in which case it will treat `x_1` as a batch
+    # therefore we sum to ensure such a thing does not happen
+    return sum(logabsdetjac(b, x_1))
+end
+
+function logabsdetjac(cl::CouplingLayer{B}, x::AbstractMatrix) where {B}
+    r = [logabsdetjac(cl, x[:, i]) for i = 1:size(x, 2)]
```

Comment:
I think we can delegate this column-wise iteration to the `logabsdetjac` function. We can then define efficient adjoints for `logabsdetjac` when `x` is `TrackedArray`. If the tracking comes from `cl`, and no scalar operations happen in `logabsdetjac`, I think `Tracker` should just work without requiring additional adjoint definitions.

Reply:
>  We can then define efficient adjoints for logabsdetjac when x is TrackedArray.
I was thinking about something similar the other week, but at the time it seemed to me like it wouldn't work. I believe my reason was just that since the only requirement is that `B` _returns_ a `Bijector`, this could have `Tracked` parts coming from somewhere else, i.e. you define `B` by closing over some `Tracked` thing. I'm uncertain how this works with defining custom adjoints :confused: 

>  I think we can delegate this column-wise iteration to the logabsdetjac function. 
I'm not entirely certain what you mean here. Do you want to make it so that `logabsdetjac` _always_ takes a matrix as input?

Reply:
I mean to define `logabsdetjac` for matrix `x`.

Reply:
So there's a slightly annoying issue.

In the case where `b = B(cl.θ(x_2))` returns a `Bijector{0}`, `logabsdetjac(b::Bijector{0}, x::AbstractMatrix)` is generally _not_ defined. Only `logabsdetjac(b::Bijector{0}, x::Union{Real, AbstractVector})` are defined.

I'm not entirely sure what to do in those cases. With the current implementation everything still works just fine. I do agree though, the current approach is sub-optimal :/

Reply:
Also, there's the issue of what to do with `B(cl.θ(x_2))`. Not all bijectors will take a `Vector` instead of `Real` as parameter and have this imply that the evaluation of this `Bijector` works on a "batch" rather than a single input. E.g. if `cl.θ(x_2)` returns a `(param_size, batch_size)`, this doesn't necessarily mean that `B(...)(x)[i] == B(cl.θ(x_2)[i, :])(x[i, :])`.

Reply:
Can't we tell this from the "dimension" of the bijector?

Reply:
Yeah, but not the "dimension" of the _parameters_ of it. It basically comes down to the fact that the following isn't true in general
```julia
# LHS might not even work
B(cl.θ(x_2))(x)[i, :] != B(cl.θ(x_2)[i, :])(x[i, :])
```

Reply:
I see, ignore my comment then. I will look into this again some other time to see if there is a way to make it work.